File: Project5/go/task1.go
/*
 * Error handling in Go
 * 
 * Svita Kiran
 * 11/13/2025
 * CS333
*/

package main

import (
    "errors"
    "fmt"
    "os"
)

func openFileDemo() {
    fmt.Println("file I/O error")

    file, err := os.Open("dne.txt")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer file.Close()

    fmt.Println("file opened")
}

func safeDivide(a, b int) (int, error) {
    if b == 0 {
        return 0, errors.New("cant divide by zero")
    }
    return a / b, nil
}

func divideDemo() {
    fmt.Println("\ndivision error")

    _, err := safeDivide(10, 0)
    if err != nil {
        fmt.Println("error:", err)
        return
    }
}

func main() {
    openFileDemo()
    divideDemo()
}

File: Project5/haskell/task1.hs
{- Error handling in Haskell

Svita Kiran
CS333
11/13/2025
-}

import System.IO
import System.IO.Error (catchIOError)
import Control.Exception (catch, SomeException)
import Text.Read (readMaybe)

fileDemo :: IO ()
fileDemo = do
    putStrLn "file I/O error"
    content <- catchIOError (readFile "dne.txt")
                            (\e -> do
                                putStrLn ("error opening file: " ++ show e)
                                return "")
    if null content
        then putStrLn "file read failed\n"
        else putStrLn ("file contents:\n" ++ content)


safeDivide :: Int -> Int -> Maybe Int
safeDivide _ 0 = Nothing
safeDivide a b = Just (a `div` b)

divideDemo :: IO ()
divideDemo = do
    putStrLn "division error"
    case safeDivide 10 0 of
        Nothing -> putStrLn "error: divide by zero\n"
        Just r  -> putStrLn ("result: " ++ show r)


main :: IO ()
main = do
    fileDemo
    divideDemo
